core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -647,12 +652,20 @@ class GroupCoordinator(val brokerId: Int,
     if (group.is(AwaitingSync))
       resetAndPropagateAssignmentError(group, Errors.REBALANCE_IN_PROGRESS)
 
-    group.transitionTo(PreparingRebalance)
-    info("Preparing to restabilize group %s with old generation %s".format(group.groupId, group.generationId))
-
-    val rebalanceTimeout = group.rebalanceTimeoutMs
-    val delayedRebalance = new DelayedJoin(this, group, rebalanceTimeout)
     val groupKey = GroupKey(group.groupId)
+
+    val (delay, timeout) = if (group.currentState != InitialRebalance) {
+      group.transitionTo(PreparingRebalance)
+      (group.rebalanceTimeoutMs, time.milliseconds() + group.rebalanceTimeoutMs)
+    } else {
+      joinPurgatory.cancelForKey(groupKey) match {
```

Comment:
@guozhangwang i don't think removing the `InitialRebalance` state is the correct thing to do. I think we'd then have to hold the state in `Empty` until we are ready to actually do the rebalance and then transition it to `PreparingRebalance`. I don't think we can just use `PreparingRebalance` as it can happen that the group goes from `Stable` -> `PreparingRebalance` and then we don't want to add the delay